rust/crypto/src/aes_gcm.rs

```diff
@@ -0,0 +1,271 @@
+        if self.msg_buf_offset > 0 {
+            for i in self.msg_buf_offset..TAG_SIZE {
+                self.msg_buf[i] = 0;
+            }
+            self.ghash.update(&self.msg_buf)?;
+        }
+
+        let mut final_block = [0u8; 16];
+        final_block[..8].copy_from_slice(&(8 * self.ad_len).to_be_bytes());
+        final_block[8..].copy_from_slice(&(8 * self.msg_len).to_be_bytes());
+
+        self.ghash.update(&final_block)?;
+        let mut hash = self.ghash.finalize()?;
+
+        for (i, b) in hash.iter_mut().enumerate() {
+            *b ^= self.nonce[i];
+        }
+
+        Ok(hash)
+    }
+}
+
+pub struct Aes256GcmEncryption {
+    ctr: Aes256Ctr32,
+    ghash: GcmGhash,
+}
+
+impl Aes256GcmEncryption {
+    fn new(
+        aes256: Aes256,
+        h: &[u8; TAG_SIZE],
+        nonce: &[u8],
+        associated_data: &[u8],
+    ) -> Result<Self> {
+        /*
+        GCM supports other sizes but 12 bytes is standard and other
+        sizes require special handling
+        */
+        if nonce.len() != NONCE_SIZE {
+            return Err(Error::InvalidNonceSize);
+        }
+
+        let mut ctr = Aes256Ctr32::new(aes256, nonce, 1)?;
+
+        let mut ghash_nonce = [0u8; 16];
+        ctr.process(&mut ghash_nonce)?;
+
+        let ghash = GcmGhash::new(&h, ghash_nonce, associated_data)?;
+        Ok(Self { ctr, ghash })
+    }
+
+    pub fn encrypt_in_place(&mut self, buf: &mut [u8]) -> Result<()> {
+        self.ctr.process(buf)?;
+        self.ghash.update(buf)?;
+        Ok(())
+    }
+
+    pub fn compute_tag(self) -> Result<[u8; TAG_SIZE]> {
+        self.ghash.finalize()
+    }
+}
+
+pub struct Aes256GcmDecryption {
+    ctr: Aes256Ctr32,
+    ghash: GcmGhash,
+}
+
+impl Aes256GcmDecryption {
+    fn new(
+        aes256: Aes256,
+        h: &[u8; TAG_SIZE],
+        nonce: &[u8],
+        associated_data: &[u8],
+    ) -> Result<Self> {
+        /*
+        GCM supports other sizes but 12 bytes is standard and other
+        sizes require special handling
+        */
+        if nonce.len() != NONCE_SIZE {
+            return Err(Error::InvalidNonceSize);
+        }
+
+        let mut ctr = Aes256Ctr32::new(aes256, nonce, 1)?;
+
+        let mut ghash_nonce = [0u8; 16];
+        ctr.process(&mut ghash_nonce)?;
+
+        let ghash = GcmGhash::new(&h, ghash_nonce, associated_data)?;
+        Ok(Self { ctr, ghash })
+    }
+
+    pub fn decrypt_in_place(&mut self, buf: &mut [u8]) -> Result<()> {
+        self.ghash.update(buf)?;
+        self.ctr.process(buf)?;
+        Ok(())
+    }
+
+    pub fn verify_tag(self, tag: &[u8]) -> Result<()> {
+        if tag.len() != TAG_SIZE {
+            return Err(Error::InvalidTag);
+        }
+
+        let computed_tag = self.ghash.finalize()?;
+
+        let tag_ok = tag.ct_eq(&computed_tag);
+
+        if tag_ok.unwrap_u8() == 0 {
+            return Err(Error::InvalidTag);
+        }
+
+        Ok(())
+    }
+}
+
+pub struct Aes256Gcm {
```

Comment:
I don't really understand what this extra type is for. I see that it has setup work, but is that work actually reused in practice? Or will it always be Aes256GcmEncryption / Aes256GcmDecryption that get reused?

Reply:
`Aes256Gcm{Encryption,Decryption}` are maintaining per-message state and are not reused. `Aes256Gcm` is caching the expanded AES key plus the encryption of all-zeros which is used across all GCM messages encrypted with that key.

Reply:
Right, but are multiple messages ever encrypted with the same key in practice? If not, it doesn't seem like an API we need to provide right now.

Reply:
Ah, I see what you mean. You're right probably we don't actually need this. Will modify